functions/index.js

```diff
@@ -1,17 +1,139 @@
+      await notifyStudentLessonPendingReview(after);
+    }
+  }
+
+  //Detect when an additional tutor offer to teach a requested lesson:
+  //more specifically when a tutor uid is added to the tutorUid list of a STUDENT_REQUESTED lesson
+  if (after.status === "STUDENT_REQUESTED" && before.tutorUid.length < after.tutorUid.length) {
+    console.log("Tutor added to the lesson's tutorUid.");
+    await notifyStudentAsTutorOfferedToTeach(after);
+  }
+});
+
+
+exports.lessonReminder = onSchedule("every 5 minutes", async (event) => {
+  const now = new Date();
+  const oneHourFromNow = new Date(now.getTime() + 60 * 60 * 1000);
+
+  try {
+    // Format the time slots to match Firestore's string format
+    const formatter = (date) =>
+      `${date.getDate().toString().padStart(2, "0")}/${
+        (date.getMonth() + 1).toString().padStart(2, "0")
+      }/${date.getFullYear()}T${date.getHours().toString().padStart(2, "0")}:${
+        date.getMinutes().toString().padStart(2, "0")
+      }:00`;
+
+    const nowString = formatter(now);
+    const oneHourString = formatter(oneHourFromNow);
+
+    console.log(`Checking lessons between ${nowString} and ${oneHourString}`);
+
+    // Query Firestore for lessons scheduled in the next hour
+    const lessonsSnapshot = await db
+      .collection("lessons")
+      .where("status", "==", "CONFIRMED")
+      .where("timeSlot", ">=", nowString)
+      .where("timeSlot", "<=", oneHourString)
+      .get();
+
+    if (lessonsSnapshot.empty) {
+      console.log("No lessons found for the next hour.");
+      return;
+    }
+
+    // Process each lesson that requires a reminder
+    const notificationPromises = lessonsSnapshot.docs.map(async (doc) => {
+      const lesson = doc.data();
+      const tutorUid = String(lesson.tutorUid).split(",")[0];
+      const studentUid = lesson.studentUid;
+
+      try {
+        // Fetch tutor and student profiles in parallel
+        const [tutorQuery, studentQuery] = await Promise.all([
+          db.collection("profiles").where("uid", "==", tutorUid).limit(1).get(),
+          db.collection("profiles").where("uid", "==", studentUid).limit(1).get(),
+        ]);
+
+        if (tutorQuery.empty || studentQuery.empty) {
+          console.error("Tutor or Student profile not found.");
+          return;
+        }
+
+        const tutorData = tutorQuery.docs[0].data();
+        const studentData = studentQuery.docs[0].data();
+
+        // Prepare notification to both tutor and student
+        const tutorPayload = preparePayload(
+          "Upcoming Lesson Reminder",
+          `Your lesson with ${studentData.firstName} starts in one hour!`,
+          tutorData.token
+        );
+
+        const studentPayload = preparePayload(
+          "Upcoming Lesson Reminder",
+          `Your lesson with ${tutorData.firstName} starts in one hour!`,
+          studentData.token
+        );
+
+        // Send notifications
+        await Promise.all([
```

Comment:
Use of Promise.all for parallel data fetching and notification sending is efficient.